File: AtDemo/SysTextViewVC/View/TableViewCell.m
//
//  TableViewCell.m
//  AtDemo
//
//  Created by XWH on 2021/4/24.
//

#import "TableViewCell.h"

#import "YYText.h"
#import "TextViewBinding.h"
#import "YYLabel.h"

@interface TableViewCell ()<UIGestureRecognizerDelegate>

@property (weak, nonatomic) IBOutlet YYLabel *yyLabel;

@end

@implementation TableViewCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    self.yyLabel.numberOfLines = 0;
    [self setupGesture];
}

- (void)setupGesture
{
    UITapGestureRecognizer *contentTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onTapReply:)];
    contentTap.delegate = self;
    [self.yyLabel addGestureRecognizer:contentTap];
}

- (void)onTapReply:(UITapGestureRecognizer *)sender {
    NSLog(@"点击了cell22");
}

#pragma mark - UIGestureRecognizerDelegate
- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
    if ([touch.view isKindOfClass:[YYLabel class]]) {
        YYLabel *label = (YYLabel *)touch.view;
        NSRange highlightRange;
        YYTextHighlight *highlight = [label _getHighlightAtPoint:[touch locationInView:label] range:&highlightRange];
        if (highlight) {
            return NO;
        }
        return YES;
    }
    return YES;
}

- (void)setModel:(DataModel *)model {
    _model = model;

    NSMutableAttributedString *muAttriSting = [[NSMutableAttributedString alloc]initWithString:model.text];
    muAttriSting.yy_font = [UIFont systemFontOfSize:25];
    muAttriSting.yy_color = UIColor.purpleColor;
    
    for (TextViewBinding *bindingModel in model.userList) {
        [muAttriSting yy_setTextHighlightRange:bindingModel.range
                                         color:UIColor.redColor
                               backgroundColor:UIColor.darkGrayColor
                                      userInfo:@{@"id":@(bindingModel.userId)}
                                     tapAction:^(UIView * _Nonnull containerView, NSAttributedString * _Nonnull text, NSRange range, CGRect rect) {
            YYLabel *yyLabel = (YYLabel *)containerView;
            YYTextHighlight *highlight = (YYTextHighlight *)[yyLabel valueForKey:@"_highlight"];
            NSDictionary *dictInfo = highlight.userInfo;
            if (dictInfo == nil) {
                return;
            }

            NSInteger toUserId = [dictInfo[@"id"] intValue];
            NSLog(@"点击了: %ld",toUserId);
        } longPressAction:nil];
    }

    self.yyLabel.attributedText = muAttriSting;
}

+ (CGFloat)rowHeightWithModel:(DataModel *)model {
    CGFloat h = LABEL_HEIGHT(model.text, [UIScreen mainScreen].bounds.size.width, 25)+5;
    return h;
}
@end

File: ATDemo(Swift)/ATDemo/TableViewCell.swift
//
//  TableViewCell.swift
//  ATDemo(Swift)
//
//  Created by XWH on 2021/5/16.
//

import Foundation
import UIKit

class TableViewCell: UITableViewCell {
    
    @IBOutlet weak var yyLabel: YYLabel!
    
    var _model: DataModel!
    public var model: DataModel! {
        set {
            _model = newValue

            let muAttriSting: NSMutableAttributedString = NSMutableAttributedString(string: newValue.text)
            muAttriSting.yy_font = UIFont.systemFont(ofSize: 25)
            muAttriSting.yy_color = UIColor.purple
            
            let arr = model.userList!
            for bindingModel in arr {
                muAttriSting.yy_setTextHighlight(NSRange(location: bindingModel.range.location, length: bindingModel.range.length), color: .red, backgroundColor: .darkGray, userInfo: ["id": bindingModel.userId]) { containerView, text, range, rect in
                    
                    let yyLabel = containerView as! YYLabel
                    let highlight = yyLabel.value(forKey: "_highlight") as! YYTextHighlight
                    let dictInfo: NSDictionary? = highlight.userInfo as NSDictionary?
                    if dictInfo == nil {
                        return
                    }
                    
                    let toUserId = dictInfo?["id"] as! Int
                    print("点击了: \(toUserId)");

                } longPressAction: { _,_,_,_ in }

            }
            yyLabel.attributedText = muAttriSting
            
        } get {
            return _model
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        yyLabel.numberOfLines = 0
        self.setupGesture()
    }
    
    func setupGesture() {
        let contentTap = UITapGestureRecognizer(target: self, action: #selector(onTapReply(_:)))
        contentTap.delegate = self
        yyLabel.addGestureRecognizer(contentTap)
    }
    
    public class func rowHeightWithModel(model: DataModel) -> CGFloat {
        let height = (model.text.heightWithFont(font: UIFont.systemFont(ofSize: 25), fixedWidth: UIScreen.main.bounds.size.width)) as CGFloat
        return height+5
    }
    
    @objc func onTapReply(_ sender: UITapGestureRecognizer) {
        print("点击了cell22")
    }
    
    override func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
        if ((touch.view?.isKind(of: YYLabel.self)) != nil) {
            let label = touch.view as! YYLabel
            var highlightRange: NSRange = NSRange(location: 0, length: 0)
            let highlight: YYTextHighlight?
                = label._getHighlight(at: touch.location(in: label), range: &highlightRange)
            if highlight != nil {
                return false
            }
            return true
        }
        return true
    }
}
